lib/reline.rb

```diff
@@ -369,89 +369,38 @@ def readline(prompt = '', add_hist = false)
       end
     end
 
-    # GNU Readline waits for "keyseq-timeout" milliseconds to see if the ESC
-    # is followed by a character, and times out and treats it as a standalone
-    # ESC if the second character does not arrive. If the second character
-    # comes before timed out, it is treated as a modifier key with the
-    # meta-property of meta-key, so that it can be distinguished from
-    # multibyte characters with the 8th bit turned on.
-    #
-    # GNU Readline will wait for the 2nd character with "keyseq-timeout"
-    # milli-seconds but wait forever after 3rd characters.
+    # GNU Readline watis for "keyseq-timeout" milliseconds when the input is
+    # ambiguous whether it is matching or matched.
+    # If the next character does not arrive within the specified timeout, input
+    # is considered as matched.
+    # `ESC` is ambiguous because it can be a standalone ESC (matched) or part of
+    # `ESC char` or part of CSI sequence (matching).
     private def read_io(keyseq_timeout, &block)
       buffer = []
+      status = :matching
       loop do
-        c = io_gate.getc(Float::INFINITY)
-        if c == -1
-          result = :unmatched
-        else
-          buffer << c
-          result = key_stroke.match_status(buffer)
-        end
-        case result
-        when :matched
-          expanded, rest_bytes = key_stroke.expand(buffer)
-          rest_bytes.reverse_each { |c| io_gate.ungetc(c) }
-          block.(expanded)
-          break
-        when :matching
-          if buffer.size == 1
-            case read_2nd_character_of_key_sequence(keyseq_timeout, buffer, c, block)
-            when :break then break
-            when :next  then next
-            end
-          end
-        when :unmatched
-          if buffer.size == 1 and c == "\e".ord
-            read_escaped_key(keyseq_timeout, c, block)
+        timeout = status == :matching_matched ? keyseq_timeout.fdiv(1000) : Float::INFINITY
+        c = io_gate.getc(timeout)
+        if c.nil? || c == -1
+          if status == :matching_matched
```

Comment:
I added a comment 61d86988d4c85532bc19caf2c81bc236498d6c14

> I assumed :matching is like partially matched and :matched is fully matched

Yes, it's correct. but `:matched` and `:matching` have overlaps. Some input will be `:matching` and in the same time it will also be `:matched`. I named it `:matching_matched`.

When three key sequence `['ab', 'abcd']` is registered, original :matched and :matching will be:
```
matched: 'ab', 'abcd'
matching: 'a', 'ab', 'abc'
overlap: 'ab'
```
The overlap part `'ab'` will be `:matching_matched`

We categorized `:matching_matched` as `:matching` before, resulted in complex implementation of read_2nd_character_of_key_sequence and read_escaped_key.

Reply:
Updated :+1: